drivers/dac/ad3552r/ad3552r.c

```diff
@@ -0,0 +1,1370 @@
+		AD3552R_MASK_DUAL_SPI_SYNCHROUNOUS_EN
+	},
+#endif
+};
+
+/* 0 -> reg addr, 1->ch0 mask, 2->ch1 mask */
+static const uint16_t addr_mask_map_ch[][3] = {
+	[AD3552R_CH_DAC_POWERDOWN] = {
+		AD3552R_REG_ADDR_POWERDOWN_CONFIG,
+		AD3552R_MASK_CH_DAC_POWERDOWN(0),
+		AD3552R_MASK_CH_DAC_POWERDOWN(1)
+	},
+	[AD3552R_CH_AMPLIFIER_POWERDOWN] = {
+		AD3552R_REG_ADDR_POWERDOWN_CONFIG,
+		AD3552R_MASK_CH_AMPLIFIER_POWERDOWN(0),
+		AD3552R_MASK_CH_AMPLIFIER_POWERDOWN(1)
+	},
+	[AD3552R_CH_OUTPUT_RANGE_SEL] = {
+		AD3552R_REG_ADDR_CH0_CH1_OUTPUT_RANGE,
+		AD3552R_MASK_CH_OUTPUT_RANGE_SEL(0),
+		AD3552R_MASK_CH_OUTPUT_RANGE_SEL(1)
+	},
+	[AD3552R_CH_TRIGGER_SOFTWARE_LDAC] = {
+		AD3552R_REG_ADDR_SW_LDAC_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+	[AD3552R_CH_HW_LDAC_MASK] = {
+		AD3552R_REG_ADDR_HW_LDAC_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+	[AD3552R_CH_SELECT] = {
+		AD3552R_REG_ADDR_CH_SELECT_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+};
+
+static const int32_t ad3542r_ch_ranges[][2] = {
+	[AD3542R_CH_OUTPUT_RANGE_0__2P5V]	= {0, 2500},
+	[AD3542R_CH_OUTPUT_RANGE_0__3V]		= {0, 3000},
+	[AD3542R_CH_OUTPUT_RANGE_0__5V]		= {0, 5000},
+	[AD3542R_CH_OUTPUT_RANGE_0__10V]	= {0, 10000},
+	[AD3542R_CH_OUTPUT_RANGE_NEG_2P5__7P5V]	= {-2500, 7500},
+	[AD3542R_CH_OUTPUT_RANGE_NEG_5__5V]	= {-5000, 5000}
+};
+
+static const int32_t ad3552r_ch_ranges[][2] = {
+	[AD3552R_CH_OUTPUT_RANGE_0__2P5V]	= {0, 2500},
+	[AD3552R_CH_OUTPUT_RANGE_0__5V]		= {0, 5000},
+	[AD3552R_CH_OUTPUT_RANGE_0__10V]	= {0, 10000},
+	[AD3552R_CH_OUTPUT_RANGE_NEG_5__5V]	= {-5000, 5000},
+	[AD3552R_CH_OUTPUT_RANGE_NEG_10__10V]	= {-10000, 10000}
+};
+
+static const int32_t gains_scaling_table[] = {
+	[AD3552R_CH_GAIN_SCALING_1]		= 1000,
+	[AD3552R_CH_GAIN_SCALING_0_5]		= 500,
+	[AD3552R_CH_GAIN_SCALING_0_25]		= 250,
+	[AD3552R_CH_GAIN_SCALING_0_125]		= 125
+};
+
+static const uint16_t ad3552r_chip_ids[] = {
+	[AD3542R_ID] = 0x4008,
+	[AD3552R_ID] = 0x4009,
+};
+
+static uint8_t _ad3552r_reg_len(uint8_t addr)
+{
+	switch (addr) {
+	case AD3552R_REG_ADDR_HW_LDAC_16B:
+	case AD3552R_REG_ADDR_CH_SELECT_16B:
+	case AD3552R_REG_ADDR_SW_LDAC_16B:
+	case AD3552R_REG_ADDR_HW_LDAC_24B:
+	case AD3552R_REG_ADDR_CH_SELECT_24B:
+	case AD3552R_REG_ADDR_SW_LDAC_24B:
+		return 1;
+	default:
+		break;
+	}
+
+	if (addr > AD3552R_REG_ADDR_HW_LDAC_24B)
+		return 3;
+	if (addr > AD3552R_REG_ADDR_HW_LDAC_16B)
+		return 2;
+
+	return 1;
+}
+
+static inline int32_t _ad3552r_get_reg_attr(struct ad3552r_desc *desc,
+		uint32_t attr,
+		uint16_t *val)
+{
+	int32_t err;
+
+	err = ad3552r_read_reg(desc, AD3552R_ATTR_REG(attr), val);
+	*val = field_get(AD3552R_ATTR_MASK(attr), *val);
+
+	return err;
+}
+
+static int32_t _ad3552r_update_reg_field(struct ad3552r_desc *desc,
+		uint8_t addr, uint32_t mask,
+		uint32_t val)
+{
+	int32_t err;
+	uint16_t reg, reg_full_mask;
+
+	reg_full_mask = (1 << (8 * _ad3552r_reg_len(addr))) - 1;
+	if (mask == reg_full_mask) {
+		reg = val;
+	} else {
```

Comment:
nitpick: this condition should not be necessary since you can just reg_write without mask.

Reply:
It is just a minor optimization.  It can be removed if seems complicated